New patient portal.

bradymiller opened the pull request:
New onsite patient portal by sjpadgett.

**Patient Onsite Portal:** Secure Chat, Secure Messaging, Patient Document Template/Import/Edit from Portal Dashboard, Profile, Lists, Appointments, Reports include ccr/ccd/selected (same as from demographic report) and chart document download, New portal patient notify button in Postings for sending invoice to patient where patient can submit to review with credit card payment and touch pad/mouse signature with document insert support. Dashboard review of profile revisions and document template handling. The end.

Rebased all work into 1 commit from prior PR:
https://github.com/openemr/openemr/pull/436

Below text is authored by bradymiller:

**Note for bradymiller**: when bring this into codebase, will need to do via commandline to ensure maintain original author of commit.

**TODO for bradymiller**:
related to the appt:
-appt time slot is defaulting to 0 (instead of 15 min for example)
-not showing the recurrent appt (the code was there to do this, but was removed)
-need to create a patient tracker record when create the appt (and flag it as being created in portal which should be easy since use the Pending status))
js translates:
-summernote.js
-summernote-ext-nugget.js
-onsiteactivityviews.js
-signer.js
-signpad.js
messaging:
-needs to incorporate messages better into OpenEMR (in counter and link from the messaging screen)
options:
-break down some typical options in globals
comments:
-remove all commented out code

bradymiller commented:
TODO:
related to the appt:
-status is defaulting to ^, and should instead be -
-appt time slot is defaulting to 0 (instead of 15 min for example)
-not showing the recurrent appt (the code was there to do this, but was removed)
-need to create a patient tracker record when create the appt (and flag it as being created in portal)
related to main portal menu:
-Signature On File link is not working
regarding upload paths:
-patients/import_templates upload to documents directory
-patients/import_templates_ui upload to documents directory
-doc_lib.php
-download_template.php
-template_menu.php
-migrate the 3 templates to documents directory
js translates:
-summernote.js
-onsiteactivityviews.js
-onsitedocuments.js
-signer.js
-signpad.js
messaging:
-whenever send a message, it always states that you are sending it to yourself
-needs to incorporate messages better into OpenEMR (in counter and link from the messaging screen)
security:
-ensure is secure and only allow access if the global is on
options:
-break down some typical options in globals
sql:
-does weird sql connection at _machine_config.php do anything. Can remove it??(if needed, then set it up correctly)
-for the zquery function, ensure sql-mode and utf8 is set up correctly
remove:
-try to remove lib/savant/tests
comments:
-remove all commented out code

sjpadgett commented:
I'm concerned you have missed some commits or working with stale version of portal. How can I branch your working copy?
As for machine_config it is required to open database so framework can write to it. Whats not setup correctly?
zquery is merely a wrapper of CCM zquery done for stand alone ccda in portal. It uses openemr sql wrappers.
The message in messaging appearing to ones self was fixed before. Use to use ip of users to determine whose who but I change it to use userid's instead. So if messages were sent from same ip, I did not know how to assign orientation. In normal use this would not be a problem but it showed in testing, thus the change.
I fixed several issues with mail like forward to pnotes and separate histories for users.

sjpadgett commented:
Signature on file not showing is weird timing issue. I fixed and will commit once I can.

bradymiller commented:
Hi Jerry,

Took that branch from the most recent code. If you have another newer
branch, just let me know, and I'll compare the 2 to ensure mine is not
behind. To work from my branch would do the following on github commandline:
git remote add bradymiller https://github.com/bradymiller/openemr.git
git fetch bradymiller
git checkout bradymiller/rebase-patient-portal_6
git checkout -b jerry-patient-portal_6

You'll then have the working branch in jerry-patient-portal_6.

Another good option here is to just keep working on your original working
branch and I'll be happy to deal with bringing all the code together
especially since it appears you may of fixed things that are not in my code.

Thanks for the quick explanation on machine_config (note those were just
notes to trigger myself to look into the connection further; didn't mean to
give impression it was wrong)

thanks,
-brady

On Wed, Mar 8, 2017 at 10:57 AM, JerryP <notifications@github.com> wrote:

> Signature on file not showing is weird timing issue. I fixed and will
> commit once I can.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/openemr/openemr/pull/474#issuecomment-285133680>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AARBuBaJofUpHhqR4M3k4RjGO2C5JOzvks5rjvofgaJpZM4L495S>
> .
>

sjpadgett commented:
I know, my communication skills seem to be lacking these days, sorry. I will go ahead and branch yours and take a look then decide on where to work. My branch is questionable so we'll see.

bradymiller commented:
Sorry on my end also. My wording on that item was definitely exaggerated in
order to capture my own attention :)
(I changed the wording to be more accurate)

On Wed, Mar 8, 2017 at 11:29 AM, JerryP <notifications@github.com> wrote:

> I know, my communication skills seem to be lacking these days, sorry. I
> will go ahead and branch yours and take a look then decide on where to
> work. My branch is questionable so we'll see.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/openemr/openemr/pull/474#issuecomment-285142906>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AARBuHyQ6VSpn5oPxV5Y0DYmA7e47GkOks5rjwGigaJpZM4L495S>
> .
>

sjpadgett commented:
Brady, I pulled your branch and only see difference where you did cleanup, translations and some sql changes but not any major differences. So I am going to pull your changes into my branch then commit new update to patient mail and continue to test using those changes. Probably best I stay out of your branch because apparently, for some reason (prob old age), I'm a real git anarchist :).
Also, looking over chat dashboard again, it is way to confusing to user and I'm going to make changes there but first I want to get C-CDA service up on pull request again with a mindful eye towards getting input on document templates M.U compliance. My thinking is to replace the older ccr/ccd download in portal with new template.

bradymiller commented:
Sounds good. The next thing I'm gonna work on is translating the js stuff, so it won't touch your plans.
And don't worry about being a git anarchist or messing up the branch; no git branch/repo is ever beyond repair :)

sjpadgett commented:
Don't worry with patient mail as I have that covered (messaging/messages.php). Also, the status symbol in appointment (^) was added for a new 'Pending' status. Not sure what (-) represents but it is meant to close in on a request/approve transaction scenario.

sjpadgett commented:
Just realized that the way new onsite portal is setup it does not allow for existing installs and those that have developed on existing portal and may wish to continue to use same. Therefore I suggest we refactor my portal to new location and we setup global for portal selection. Kind of important.

bradymiller commented:
Thanks for the Pending ^ pointer (will add this new element to the appt statuses list).

That is a great idea to migrate the portal to another directory. This would then also allow us to commit the portal into the main codebase much earlier (since the prior portal would still be functional), which then would make the separate git/github issues go away. Let me know if you need help on this; I'd be glad to take this on.

sjpadgett commented:
Okay and thanks because right now I am heavy into modifying patient mail
and need to get ccda service up. So if you could wait a day I will try to
get my modifications on mail up today then turn to ccda while you do
refactor.

On Fri, Mar 10, 2017 at 12:35 PM, Brady Miller <notifications@github.com>
wrote:

> Thanks for the Pending ^ pointer (will add this new element to the appt
> statuses list).
>
> That is a great idea to migrate the portal to another directory. This
> would then also allow us to commit the portal into the main codebase much
> earlier (since the prior portal would still be functional), which then
> would make the separate git/github issues go away. Let me know if you need
> help on this; I'd be glad to take this on.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/openemr/openemr/pull/474#issuecomment-285733289>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ABxhkQ1Y21rSHXFJgJiqE1pAKktb1_gsks5rkYnKgaJpZM4L495S>
> .
>



-- 
Jerry Padgett
(Skype) 813-699-4264
(Google Voice) 813-644-2333
Padgett Systems and Consulting Brandon, Florida 33511

bradymiller commented:
Sounds good. just let me know when then patient mail code is available and then I'll take the baton on this.

sjpadgett commented:
Here is my tree based on your branch:
https://github.com/sjpadgett/openemr/tree/jerry-patient-portal_6
Still have more I want to do to mail but can wait till you get refactored and I'll work ccda service for awhile.

bradymiller commented:
got it working.
I did the following:
1. Migrated to portal directory
2. Rebased everything into 1 commit
3. Updated to the most recent codebase.
4. Rebased this pull request with this.

Now that this is separate, plan to place this in the codebase soon (once ensure no security issues). I plan to keep chipping away at the TODO list.

-brady

sjpadgett commented:
Concerning your list.
-migrate the 3 templates to documents directory: 
 Because the templates are shared and generic I was not sure Document directory is right place for them because once template is processed by dashboard the reviewed is saved to document directory. Still simple matter to move.
-when send a message, it creates a duplicate message:
 Meant to be this way. Each user needs to have their own copy of messages for disposal, status and mostly privacy. This is setup to emulate an email client/server scenario where each user essentially has a mail box. I could setup to share same message and have different status and delete fields to manage disposal but that defeats message ownership and possibly raise issues concerning privacy. I've been back and forth on this many times and am apprehensive either way.

sjpadgett commented:
Brady, Pulled the Savant test scripts. You probably see but my branch is now: https://github.com/sjpadgett/openemr/tree/portal-moved-rebased
- How would you like to see the directory structure for the patient document pending templates?
I am doing some quick changes to chat now.

bradymiller commented:
Thanks!
Would put all the patient document templates and related stuff in documents/onsite_portal_documents/...
(feel free to create subdir on this as needed)
The reason use documents is then don't need to deal with giving write access every time create a new directory  (done already) and it won't break folks backup mechanisms.

bradymiller commented:
btw, the branch looks really nice (we may need to rethink your git anarchist reputation :) )

sjpadgett commented:
Ya really just paid yourself a compliment because it's just a recycle of
your branch. Figured I couldn't go wrong there. :)

On Mon, Mar 13, 2017 at 3:49 PM, Brady Miller <notifications@github.com>
wrote:

> btw, the branch looks really nice (we may need to rethink your git
> anarchist reputation :) )
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/openemr/openemr/pull/474#issuecomment-286222594>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ABxhkeHkuNPXsIap2ws-nvGawRGIuk17ks5rlZ2zgaJpZM4L495S>
> .
>



-- 
Jerry Padgett
(Skype) 813-699-4264
(Google Voice) 813-644-2333
Padgett Systems and Consulting Brandon, Florida 33511

bradymiller commented on portal/patient/_machine_config.php:
```diff
@@ -0,0 +1,87 @@
+GlobalConfig::$CONNECTION_SETTING->Username = $GLOBALS['login'];
+GlobalConfig::$CONNECTION_SETTING->Password = $GLOBALS['pass'];
+GlobalConfig::$CONNECTION_SETTING->Type = "MySQLi";
+GlobalConfig::$CONNECTION_SETTING->Charset = "utf8";
```
Should include above line based on teh following flag:
https://github.com/openemr/openemr/blob/master/sites/default/sqlconf.php#L7

bradymiller commented on portal/patient/fwk/libs/verysimple/DB/DataDriver/MySQLi.php:
```diff
@@ -0,0 +1,234 @@
+		}
+		
+		if ($charset) {
+			mysqli_set_charset ( $connection, $charset );
```
TODO:
I need to look at this function to ensure it's doing SET NAMES UTF8
(github search not working, so will do better searching at home)

bradymiller commented on portal/patient/_machine_config.php:
```diff
@@ -0,0 +1,87 @@
+GlobalConfig::$CONNECTION_SETTING->Password = $GLOBALS['pass'];
+GlobalConfig::$CONNECTION_SETTING->Type = "MySQLi";
+GlobalConfig::$CONNECTION_SETTING->Charset = "utf8";
+GlobalConfig::$CONNECTION_SETTING->Multibyte = true;
```
Need to blank out sql-mode after every connection(query is "SET sql_mode = ''"). It does not look like this scheme offers a way to send a sql query every time open a connection without editing portal/patient/fwk/libs/verysimple/DB/DataDriver/MySQLi.php . @sjpadgett , is there a place that you know of to do this type of thing?

sjpadgett commented:
I haven't visited these libraries in a while but I see your point. On the
short term I may be able to intercept sql calls at the controller and
bypass the library much like ZH does in CCM or maybe use reporter
controller beings I am not changing the model dynamically. Or run routing
off backbone and lose phreeze (much more effort). I'll look at controller
solution now and let you know.

On Mon, Mar 13, 2017 at 4:11 PM, Brady Miller <notifications@github.com>
wrote:

> *@bradymiller* commented on this pull request.
> ------------------------------
>
> In portal/patient/_machine_config.php
> <https://github.com/openemr/openemr/pull/474#discussion_r105759343>:
>
> > +	}
> +
> +require_once 'verysimple/Phreeze/ConnectionSetting.php';
> +require_once ( "verysimple/HTTP/RequestUtil.php" );
> +
> +/**
> + * database connection settings
> + */
> +GlobalConfig::$CONNECTION_SETTING = new ConnectionSetting();
> +GlobalConfig::$CONNECTION_SETTING->ConnectionString = $GLOBALS['host'] . ":" . $GLOBALS['port'];
> +GlobalConfig::$CONNECTION_SETTING->DBName = $GLOBALS['dbase'];
> +GlobalConfig::$CONNECTION_SETTING->Username = $GLOBALS['login'];
> +GlobalConfig::$CONNECTION_SETTING->Password = $GLOBALS['pass'];
> +GlobalConfig::$CONNECTION_SETTING->Type = "MySQLi";
> +GlobalConfig::$CONNECTION_SETTING->Charset = "utf8";
> +GlobalConfig::$CONNECTION_SETTING->Multibyte = true;
>
> Need to blank out sql-mode after every connection(query is "SET sql_mode =
> ''"). It does not look like this scheme offers a way to send a sql query
> every time open a connection without editing portal/patient/fwk/libs/
> verysimple/DB/DataDriver/MySQLi.php . @sjpadgett
> <https://github.com/sjpadgett> , is there a place that you know of to do
> this type of thing?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openemr/openemr/pull/474#pullrequestreview-26654116>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABxhkcDjKJv61D65Ibyll9UyyT7P3t4_ks5rlaL2gaJpZM4L495S>
> .
>



-- 
Jerry Padgett
(Skype) 813-699-4264
(Google Voice) 813-644-2333
Padgett Systems and Consulting Brandon, Florida 33511

bradymiller commented on portal/lib/appsql.class.php:
```diff
@@ -0,0 +1,351 @@
+     *            Error Display True / False
+     * @return type
+     */
+    public function zQuery( $sql, $params = '', $log = FALSE, $error = TRUE ){
```
This zQuery and log mechanism is good to go.

sjpadgett commented:
That's assuming we want to get rid of any framework driver and only use ado wrappers otherwise I say we just mod the library because at some point I'm going to write it away, anyway.

bradymiller commented:
modding the library sounds good. I'll do this and make it clear that we did it. I'll also bring in the utf8 flag to the main encoding connection setting. After this and the document path stuff is adjusted, then will bring it into the codebase and work on it from there.

sjpadgett commented:
Okay, I'll have the path changes soon. Thanks

bradymiller commented:
@sjpadgett , I just merged in your commit deleting savant scripts. Just keep working on your branch and I'll bring stuff over. I'm gonna start on the database stuff.

bradymiller commented:
I just committed the database changes. Turns out was very minimal and didn't have to modify the main library:
https://github.com/openemr/openemr/pull/474/commits/87ee47b19e8793cd3d88d3915ebd74307646d5c4
(btw, just keep working on your branch and don't worry about this code)

sjpadgett commented:
Kind of obvious when you see it, clever.
"(btw, just keep working on your branch and don't worry about this code)" : Not sure how to take that :)

bradymiller commented:
yep, I feel a bit foolish with my hoopla of modifying the main library when the BootstrapSQL thing was right there in front of me :)

sjpadgett commented:
I looked at it three times and didn't get it, seems I tend to look for the
complicated first and miss the obvious. I found some bugs in process of
moving documents so I'm fixing those and will have done soon. Also, seems
the demographics patient document upload/move etc is not working as
expected. Have you tested this lately? I kind of let go for moment and will
look at closer after portal docs. Another note, I spent 5 hours last night
tracking down a billing claims problem for my clinic (v4.2.2) where primary
insurance was changed after encounter was created and claim would only
allow old insurance to be selected (apparently in version 4.2.2 this is
intended based on visit date), appears that is gone in v5.0.1 but, that is
not what troubled me too much, it is the fact that once a claim is created
and payer is set in claim, fee sheet will overwrite all payer versions
history after any changes. Bottom line, should not, once a claim is
generated, any payer info for a claim stay with the version the claim is
generated against and should never be changed? Any payer or fee sheet
changes will be reflected in new version. I see this as a possible medicare
audit issue.

On Tue, Mar 14, 2017 at 1:57 PM, Brady Miller <notifications@github.com>
wrote:

> yep, I feel a bit foolish with my hoopla of modifying the main library
> when the BootstrapSQL thing was right there in front of me :)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openemr/openemr/pull/474#issuecomment-286506917>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ABxhkVyAyvxKBYMcSBKftqpqOtAGueNQks5rltT5gaJpZM4L495S>
> .
>



-- 
Jerry Padgett
(Skype) 813-699-4264
(Google Voice) 813-644-2333
Padgett Systems and Consulting Brandon, Florida 33511

sjpadgett commented:
I always intended to add popup reminder if documents/forms are available for patient to fill out such as hipaa/privacy agreements etc when they log in to portal, also a right click context popup for inserting tags/directives within template editor in dashboard; should I go ahead and do that now or after put in code base?
I don't see any reason the template editor can't also be used to edit demographics patient document templates as well.

sjpadgett commented:
Okay, Patient Documents templates move and refactor is done. I'm actually beginning to like this module, it has some potential.
I'm still going to add a help or right click insert for template tags in editor otherwise folks won't know what to use.

bradymiller commented:
@sjpadgett ,
Just brought in your 3 commits. This is very exciting; very close to getting this into the main codebase (only things left are 2 database footprint items; a pending appt status item and checking of the categories algorithm). I'm happy to do this; just will need a good night sleep tonight before going through the categories numbering algorithm tomorrow since that can turn the brain into jello pretty quickly.

Agree the billing issue is odd. As I recall, @sunsetsystems  made it so that the 'Effective Date' date is what decides whom the payer is (so if change primary payer and the effective date, then the prior bills should be still mapped to the prior primary payer by using the effective date); hopefully that mechanism is not broken. Maybe @sunsetsystems has some input on this.

Regarding sequencing of code, follow your hearts content on that one :) This portal is going into the codebase in the very near future and happy to bring in your code changes either before or after that. They all sound like great features, although the sounds of ccda are beckoning us :)
-brady

bradymiller commented:
Almost forgot, I played around with the patient summary screen->Documents and was able to upload/move things ok. There have been several changes to this code recently so a bug is definitely possible. Any specific steps to recreate the problem?

sjpadgett commented:
Nope, sorry, my problem was I deleted file in directory but document still
resides in table. Actually really nice, only thing I would like to do at
some point is give a button to pop out the document to view better.

On Wed, Mar 15, 2017 at 6:25 AM, Brady Miller <notifications@github.com>
wrote:

> Almost forgot, I played around with the patient summary screen->Documents
> and was able to upload/move things ok. There have been several changes to
> this code recently so a bug is definitely possible. Any specific steps to
> recreate the problem?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openemr/openemr/pull/474#issuecomment-286700972>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ABxhkev_3mItu25R8Y0X5OEa9e0u98HYks5rl7ybgaJpZM4L495S>
> .
>



-- 
Jerry Padgett
(Skype) 813-699-4264
(Google Voice) 813-644-2333
Padgett Systems and Consulting Brandon, Florida 33511

sjpadgett commented:
I couldn't help myself and had to add drop down button to select available tags to insert in template at cursor for the import template editor.
Going back to ccda and leaving this till later.

bradymiller commented:
@sjpadgett , just brought in your tags/directive work into the PR. gonna now work on the above discussed database stuff. if all goes well, there many be a shiny new patient portal in the codebase by daybreak.

bradymiller commented:
@sjpadgett 
I added a commit with minor fixes.
I then did a rebase of it all into 1 commit and brought this all into the main codebase. Congratulations and thank you for the awesome contribution!
I'm gonna close this PR and start an Issues ticket where will I will track the TODO list.

bradymiller commented:
btw,

Added some basic documentation on the wiki:
http://www.open-emr.org/wiki/index.php/Patient_Portal#Version_2_Native_Patient_Portal

And announced this new feature on the forums:
https://community.open-emr.org/t/new-patient-portal-has-been-brought-into-the-development-codebase/8223

And it's working on the development demo:
http://www.open-emr.org/wiki/index.php/Development_Demo#Daily_Build_Development_Demo
(note am working on a issue in the development demo to get it to work fully)

sunsetsystems commented:
Not sure I understand the billing issue. But insurer history should definitely be preserved in the insurance_data table, and the effective date is key to matching the correct entry to a given encounter.

sjpadgett commented:
Mute point I guess because claims version history and payer changes in
version 5.0.1 work as expected and while I've not tested version 5.0.0 I
would be pretty sure it is okay there as well. However, in version 4.2.2
fee sheet updates wipes out payer history in claims if insurance was
changed and claim re-opened for some reason. This issue came from having an
encounter w/DOS and insurance payer was changed between time of visit and
claim generation. Couldn't generate claim against new payer due to
effective date after DOS. Not sure if this happens often and a need for
override or not but, my main concern that payer history is maintained via
claim version history is satisfied to work correctly in new release. Thanks
for follow up.

On Thu, Mar 16, 2017 at 2:06 PM, Rod Roark <notifications@github.com> wrote:

> Not sure I understand the billing issue. But insurer history should
> definitely be preserved in the insurance_data table, and the effective date
> is key to matching the correct entry to a given encounter.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openemr/openemr/pull/474#issuecomment-287143651>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ABxhkZ6kELPYvUWcsKDkYrNqtBi39lDgks5rmXoqgaJpZM4L495S>
> .
>



-- 
Jerry Padgett
(Skype) 813-699-4264
(Google Voice) 813-644-2333
Padgett Systems and Consulting Brandon, Florida 33511

sunsetsystems commented:
OK. Note when you enter a new payer it's important to get the effective date right and it does not have to be the data entry date. Also at that time, any existing payers with a later effective date will be wiped out and will have to be re-entered if needed; that's the downside of not having a full GUI for managing multiple payers.

sjpadgett commented:
Noted and shoot golly gee wiz, I completely missed that you can change
effective date:) Thanks for correcting me. Now I see why the attempt in
v4.2.2 to maintain old insurance payer information in insurance data. It
was almost there.

On Thu, Mar 16, 2017 at 3:59 PM, Rod Roark <notifications@github.com> wrote:

> OK. Note when you enter a new payer it's important to get the effective
> date right and it does not have to be the data entry date. Also at that
> time, any existing payers with a later effective date will be wiped out and
> will have to be re-entered if needed; that's the downside of not having a
> full GUI for managing multiple payers.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openemr/openemr/pull/474#issuecomment-287173944>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ABxhkWQb7yyPWOIzzq5gPeOuZCUQTodWks5rmZSigaJpZM4L495S>
> .
>



-- 
Jerry Padgett
(Skype) 813-699-4264
(Google Voice) 813-644-2333
Padgett Systems and Consulting Brandon, Florida 33511
